pkg/rpc/client/wsclient.go

```diff
@@ -233,33 +233,39 @@ func (c *WSClient) wsWriter() {
 	}
 }
 
-func (c *WSClient) registerRespChannel(id int, ch chan *response.Raw) {
+func (c *WSClient) addResponse(id int, resp *response.Raw) {
 	c.respLock.Lock()
 	defer c.respLock.Unlock()
-	c.respChannels[id] = ch
+
+	c.responses[id] = resp
 }
 
-func (c *WSClient) getResponseChannel(id int) chan *response.Raw {
-	c.respLock.RLock()
-	defer c.respLock.RUnlock()
-	return c.respChannels[id]
+func (c *WSClient) popResponse(id int) (*response.Raw, error) {
+	c.respLock.Lock()
+	defer c.respLock.Unlock()
+
+	if resp, ok := c.responses[id]; ok {
+		delete(c.responses, id)
+		return resp, nil
+	}
+	return nil, fmt.Errorf("failed to retrieve response by ID %d", id)
 }
 
 func (c *WSClient) makeWsRequest(r *request.Raw) (*response.Raw, error) {
-	ch := make(chan *response.Raw)
-	c.registerRespChannel(r.ID, ch)
-
 	select {
 	case <-c.done:
 		return nil, errors.New("connection lost before sending the request")
 	case c.requests <- r:
 	}
-	select {
-	case <-c.done:
-		return nil, errors.New("connection lost while waiting for the response")
-	case resp := <-ch:
-		close(ch)
-		return resp, nil
+	for {
+		select {
+		case <-c.done:
+			return nil, errors.New("connection lost while waiting for the response")
+		case id := <-c.responseNotifyCh:
```

Comment:
_Any_ requester can receive this event which will only be emitted once into the `c.responseNotifyCh`. So this just won't work. Creating a channel for a request is OK here, it's nothing compared to the inevitable networking overhead.